arangod/RocksDBEngine/RocksDBRecoveryManager.cpp

```diff
@@ -375,49 +385,76 @@ class WBReader final : public rocksdb::WriteBatch::Handler {
-    return false;
-  }
+    // Tell the WriteBatch reader the transaction markers to look for
+    handler = std::make_unique<WBReader>(engine->settingsManager()->counterSeqs());
 
-  while (iterator->Valid()) {
-    s = iterator->status();
-    if (s.ok()) {
-      rocksdb::BatchResult batch = iterator->GetBatch();
-      handler->currentSeqNum = batch.sequence;
-      s = batch.writeBatchPtr->Iterate(handler.get());
-    }
-    if (!s.ok()) {
-      LOG_TOPIC(ERR, Logger::ENGINES) << "error during WAL scan";
-      break;
+    auto minTick = std::min(engine->settingsManager()->earliestSeqNeeded(),
+                            engine->releasedTick());
+    std::unique_ptr<rocksdb::TransactionLogIterator> iterator;  // reader();
+    rocksdb::Status s = _db->GetUpdatesSince(
+        minTick, &iterator, rocksdb::TransactionLogIterator::ReadOptions(true));
+
+    rv = rocksutils::convertStatus(s);
+
+    if(rv.ok()){
+      while (iterator->Valid()) {
+        s = iterator->status();
+        if (s.ok()) {
+          rocksdb::BatchResult batch = iterator->GetBatch();
+          handler->currentSeqNum = batch.sequence;
+          s = batch.writeBatchPtr->Iterate(handler.get());
+        }
+
+        std::string const msg = "error during WAL scan ";
+
+        if (!s.ok()) {
+          LOG_TOPIC(ERR, Logger::ENGINES) << msg;
+          rv = rocksutils::convertStatus(s);
+          rv.reset(rv.errorNumber(), msg + rv.errorMessage());
+          break;
+        }
+
+        iterator->Next();
+      }
+
+      if(rv.ok()){
+        LOG_TOPIC(TRACE, Logger::ENGINES)
+            << "finished WAL scan with " << handler->deltas.size();
+        for (std::pair<uint64_t, RocksDBSettingsManager::CounterAdjustment> pair :
+             handler->deltas) {
+          engine->settingsManager()->updateCounter(pair.first, pair.second);
+          LOG_TOPIC(TRACE, Logger::ENGINES)
+              << "WAL recovered " << pair.second.added() << " PUTs and "
+              << pair.second.removed() << " DELETEs for objectID " << pair.first;
+        }
+
+        if(!(handler->deltas.size() > 0)){
+          rv.reset(TRI_ERROR_INTERNAL, "error in wal recovery - deltas size less equal than 0");
+        }
+      }
     }
+  } CATCH_TO_RESULT(rv,TRI_ERROR_INTERNAL);
 
-    iterator->Next();
-  }
+  auto shutdownRv = handler->shutdownWBReader();
 
-  LOG_TOPIC(TRACE, Logger::ENGINES)
-      << "finished WAL scan with " << handler->deltas.size();
-  for (std::pair<uint64_t, RocksDBSettingsManager::CounterAdjustment> pair :
-       handler->deltas) {
-    engine->settingsManager()->updateCounter(pair.first, pair.second);
-    LOG_TOPIC(TRACE, Logger::ENGINES)
-        << "WAL recovered " << pair.second.added() << " PUTs and "
-        << pair.second.removed() << " DELETEs for objectID " << pair.first;
+  if(rv.ok()) {
+    rv = std::move(shutdownRv);
+  } else {
+    if(shutdownRv.fail()){
+      rv.reset(rv.errorNumber(), rv.errorMessage() + shutdownRv.errorMessage());
```

Comment:
Should there be some separator character between `rv.errorMessage()` and `shutdownRv.errorMessage()`?
Otherwise the two messages are just glued together without any whitespace between, aren't they?